# LAB | Ensemble Methods

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In this Lab, you should try different ensemble methods in order to see if can obtain a better model than before. In order to do a fair comparison, you should perform the same feature scaling, engineering applied in previous Lab.

In [1]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import BaggingRegressor, RandomForestRegressor,AdaBoostRegressor, GradientBoostingRegressor

from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import BaggingRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [2]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

Now perform the same as before:
- Feature Scaling
- Feature Selection


In [3]:
#drop missing values
df_cleaned = spaceship.dropna()
df_cleaned.shape

(6606, 14)

In [4]:
#format cabin values
df_cleaned['Cabin'] = df_cleaned['Cabin'].str[0]
df_cleaned['Cabin'].value_counts()

/var/folders/76/0pntsdld7bg_p5gzgms4gghw0000gn/T/ipykernel_80821/3361844794.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['Cabin'] = df_cleaned['Cabin'].str[0]


Cabin
F    2152
G    1973
E     683
B     628
C     587
D     374
A     207
T       2
Name: count, dtype: int64

In [5]:
#Drop PassengerId and Name
df_cleaned2 = df_cleaned.drop(columns=['PassengerId','Name'])

In [6]:
#feature encoding.
# One-hot encoding for sex and title
non_numerical_cols = df_cleaned2.select_dtypes(include=['object', 'category']).columns
df_cleaned3 = pd.get_dummies(df_cleaned2, columns=non_numerical_cols)

df_cleaned3.head()

Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck  Transported  \
0  39.0          0.0        0.0           0.0     0.0     0.0        False   
1  24.0        109.0        9.0          25.0   549.0    44.0         True   
2  58.0         43.0     3576.0           0.0  6715.0    49.0        False   
3  33.0          0.0     1283.0         371.0  3329.0   193.0        False   
4  16.0        303.0       70.0         151.0   565.0     2.0         True   

   HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  ...  Cabin_D  \
0             False               True            False  ...    False   
1              True              False            False  ...    False   
2             False               True            False  ...    False   
3             False               True            False  ...    False   
4              True              False            False  ...    False   

   Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0    False    False    False    False                    False   
1    False     True    False    False                    False   
2    False    False    False    False                    False   
3    False    False    False    False                    False   
4    False     True    False    False                    False   

   Destination_PSO J318.5-22  Destination_TRAPPIST-1e  VIP_False  VIP_True  
0                      False                     True       True     False  
1                      False                     True       True     False  
2                      False                     True      False      True  
3                      False                     True       True     False  
4                      False                     True       True     False  

[5 rows x 25 columns]

### Select features and Target

In [7]:
# Drop the 'Transported' column
features = df_cleaned3.drop(columns=['Transported'])

# Display the first few rows of the resulting DataFrame
features.head()

Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck  \
0  39.0          0.0        0.0           0.0     0.0     0.0   
1  24.0        109.0        9.0          25.0   549.0    44.0   
2  58.0         43.0     3576.0           0.0  6715.0    49.0   
3  33.0          0.0     1283.0         371.0  3329.0   193.0   
4  16.0        303.0       70.0         151.0   565.0     2.0   

   HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  CryoSleep_False  ...  \
0             False               True            False             True  ...   
1              True              False            False             True  ...   
2             False               True            False             True  ...   
3             False               True            False             True  ...   
4              True              False            False             True  ...   

   Cabin_D  Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0    False    False    False    False    False                    False   
1    False    False     True    False    False                    False   
2    False    False    False    False    False                    False   
3    False    False    False    False    False                    False   
4    False    False     True    False    False                    False   

   Destination_PSO J318.5-22  Destination_TRAPPIST-1e  VIP_False  VIP_True  
0                      False                     True       True     False  
1                      False                     True       True     False  
2                      False                     True      False      True  
3                      False                     True       True     False  
4                      False                     True       True     False  

[5 rows x 24 columns]

In [8]:
# Select the 'Transported' column
target = df_cleaned3[['Transported']]

# Display the first few rows of the 'target' DataFrame
target.head()

Transported
0        False
1         True
2        False
3        False
4         True

### Perform Train Test Split

In [9]:
#Perform Train Test Split
from sklearn.model_selection import train_test_split


X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.20, random_state=0)

In [10]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape 

((5284, 24), (1322, 24), (5284, 1), (1322, 1))

In [11]:
#create an instance of the normalizer
from sklearn.preprocessing import MinMaxScaler
normalizer = MinMaxScaler()
normalizer.fit(X_train)

MinMaxScaler()

In [12]:
X_train_norm = normalizer.transform(X_train)
X_test_norm = normalizer.transform(X_test)

In [13]:
X_train_norm = pd.DataFrame(X_train_norm, columns = X_train.columns)
X_train_norm.head()

Age  RoomService  FoodCourt  ShoppingMall       Spa    VRDeck  \
0  0.405063      0.00000   0.000000      0.000000  0.000000  0.000000   
1  0.050633      0.00000   0.000000      0.000000  0.000000  0.000000   
2  0.379747      0.00000   0.007916      0.000000  0.051276  0.000000   
3  0.215190      0.00131   0.000000      0.046111  0.016378  0.000049   
4  0.329114      0.00000   0.000000      0.000000  0.000000  0.000000   

   HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  CryoSleep_False  ...  \
0               0.0                0.0              1.0              0.0  ...   
1               1.0                0.0              0.0              0.0  ...   
2               1.0                0.0              0.0              1.0  ...   
3               0.0                0.0              1.0              1.0  ...   
4               1.0                0.0              0.0              0.0  ...   

   Cabin_D  Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0      0.0      1.0      0.0      0.0      0.0                      0.0   
1      0.0      0.0      0.0      1.0      0.0                      0.0   
2      0.0      0.0      1.0      0.0      0.0                      0.0   
3      0.0      1.0      0.0      0.0      0.0                      0.0   
4      0.0      0.0      0.0      1.0      0.0                      0.0   

   Destination_PSO J318.5-22  Destination_TRAPPIST-1e  VIP_False  VIP_True  
0                        0.0                      1.0        1.0       0.0  
1                        0.0                      1.0        1.0       0.0  
2                        0.0                      1.0        1.0       0.0  
3                        0.0                      1.0        1.0       0.0  
4                        0.0                      1.0        1.0       0.0  

[5 rows x 24 columns]

In [14]:
X_test_norm = pd.DataFrame(X_test_norm, columns = X_test.columns)
X_test_norm.head()

Age  RoomService  FoodCourt  ShoppingMall      Spa  VRDeck  \
0  0.632911          0.0   0.000000           0.0  0.00000     0.0   
1  0.227848          0.0   0.000000           0.0  0.00000     0.0   
2  0.189873          0.0   0.000000           0.0  0.00000     0.0   
3  0.658228          0.0   0.000000           0.0  0.00000     0.0   
4  0.784810          0.0   0.054775           0.0  0.07774     0.0   

   HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  CryoSleep_False  ...  \
0               0.0                0.0              1.0              0.0  ...   
1               1.0                0.0              0.0              0.0  ...   
2               1.0                0.0              0.0              0.0  ...   
3               1.0                0.0              0.0              0.0  ...   
4               0.0                1.0              0.0              1.0  ...   

   Cabin_D  Cabin_E  Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0      0.0      0.0      1.0      0.0      0.0                      1.0   
1      0.0      0.0      0.0      1.0      0.0                      0.0   
2      0.0      0.0      0.0      1.0      0.0                      1.0   
3      0.0      0.0      0.0      1.0      0.0                      0.0   
4      0.0      0.0      0.0      0.0      0.0                      0.0   

   Destination_PSO J318.5-22  Destination_TRAPPIST-1e  VIP_False  VIP_True  
0                        0.0                      0.0        1.0       0.0  
1                        1.0                      0.0        1.0       0.0  
2                        0.0                      0.0        1.0       0.0  
3                        1.0                      0.0        1.0       0.0  
4                        0.0                      1.0        0.0       1.0  

[5 rows x 24 columns]

**Model Selection** - now you will try to apply different ensemble methods in order to get a better model

### Bagging and Pasting

In [15]:
# Check the number of samples in X_train_norm
n_samples = X_train_norm.shape[0]
print(f"Number of samples in training data: {n_samples}")

Number of samples in training data: 5284


In [16]:
# Initialize the Bagging Regressor
bagging_reg = BaggingRegressor(DecisionTreeRegressor(max_depth=20),
                               n_estimators=100,
                               max_samples = 5284)

In [17]:
# Training the Bagging model with normalized data
bagging_reg.fit(X_train_norm, y_train)

/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/sklearn/ensemble/_bagging.py:508: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return column_or_1d(y, warn=True)


BaggingRegressor(estimator=DecisionTreeRegressor(max_depth=20),
                 max_samples=5284, n_estimators=100)

In [18]:
# evaluate model's performance
pred = bagging_reg.predict(X_test_norm)

print("MAE", mean_absolute_error(pred, y_test))
print("RMSE", mean_squared_error(pred, y_test, squared=False))
print("R2 score", bagging_reg.score(X_test_norm, y_test))

MAE 0.2698346991172517
RMSE 0.38331092958896557
R2 score 0.4122909250305723


In [19]:
#with  max_samples=1000,
# Initialize the Pasting model
pasting_reg = BaggingRegressor(
  estimator=DecisionTreeRegressor(max_depth=20),
  n_estimators=100,
  max_samples=1000,
  bootstrap=False  # This ensures that no bootstrap is done, which is characteristic of Pasting
)

# Train the model with normalized data
pasting_reg.fit(X_train_norm, y_train)

# Evaluate the model's performance
pred = pasting_reg.predict(X_test_norm)

print("MAE:", mean_absolute_error(pred, y_test))
print("RMSE:", mean_squared_error(pred, y_test, squared=False))
print("R2 score:", pasting_reg.score(X_test_norm, y_test))

/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/sklearn/ensemble/_bagging.py:508: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return column_or_1d(y, warn=True)


MAE: 0.2772030575264917
RMSE: 0.3812646549204892
R2 score: 0.418549051633441


### Random Forests (Random Patches/Forest)

In [20]:
#Initialize a Random Forest
forest = RandomForestRegressor(n_estimators=100,
                             max_depth=20)

In [21]:
#Training the model
forest.fit(X_train_norm, y_train)

/var/folders/76/0pntsdld7bg_p5gzgms4gghw0000gn/T/ipykernel_80821/3171514955.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  forest.fit(X_train_norm, y_train)


RandomForestRegressor(max_depth=20)

In [22]:
#Evaluate the model
pred = forest.predict(X_test_norm)

print("MAE", mean_absolute_error(pred, y_test))
print("RMSE", mean_squared_error(pred, y_test, squared=False))
print("R2 score", forest.score(X_test_norm, y_test))

MAE 0.2692586427516191
RMSE 0.38446792832611654
R2 score 0.40873764835449644


### Gradient Boosting

In [23]:
#Initialize a AdaBoost model
gb_reg = GradientBoostingRegressor(max_depth=20,
                                   n_estimators=100)

In [24]:
#Training the model
gb_reg.fit(X_train_norm, y_train)

/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/sklearn/ensemble/_gb.py:437: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


GradientBoostingRegressor(max_depth=20)

In [25]:
#Evaluate the model
pred = gb_reg.predict(X_test_norm)

print("MAE", mean_absolute_error(pred, y_test))
print("RMSE", mean_squared_error(pred, y_test, squared=False))
print("R2 score", gb_reg.score(X_test_norm, y_test))

MAE 0.25872055038988434
RMSE 0.4222125519620773
R2 score 0.28694624386268064


### Adaptive Boosting

In [26]:
# Initialize a AdaBoost model
ada_reg = AdaBoostRegressor(DecisionTreeRegressor(max_depth=20),
                            n_estimators=100)

In [27]:
#Training the model
ada_reg.fit(X_train_norm, y_train)

/opt/anaconda3/envs/myenv/lib/python3.8/site-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


AdaBoostRegressor(estimator=DecisionTreeRegressor(max_depth=20),
                  n_estimators=100)

In [28]:
#Evaluate the model
pred = ada_reg.predict(X_test_norm)

print("MAE", mean_absolute_error(pred, y_test))
print("RMSE", mean_squared_error(pred, y_test, squared=False))
print("R2 score", ada_reg.score(X_test_norm, y_test))

MAE 0.24934278910432295
RMSE 0.4306510505997418
R2 score 0.25815869046935447


Which model is the best and why?

#### Bagging and Pasting max_samples = 5284
- MAE 0.2698346991172517
- RMSE 0.38331092958896557
- R2 score 0.4122909250305723
  
  
#### Bagging and Pasting max_samples = 1000
- MAE: 0.2772030575264917
- RMSE: 0.3812646549204892
- R2 score: 0.418549051633441


#### Random Forests (Random Patches/Forest)
- MAE 0.2692586427516191
- RMSE 0.38446792832611654
- R2 score 0.40873764835449644


#### Gradient Boosting
- MAE 0.25872055038988434
- RMSE 0.4222125519620773
- R2 score 0.28694624386268064


#### Adaptive Boosting
- MAE 0.24934278910432295
- RMSE 0.4306510505997418
- R2 score 0.25815869046935447

### Conclusion: 
Considering both the RMSE and R2 score, the Bagging and Pasting (max_samples = 1000) model appears to perform the best 

- RMSE: Lower is better, and the Bagging and Pasting model has a lower RMSE.
- R2 Score: Higher is better, and the Bagging and Pasting model has a higher R2 score.